test/settings/settings_test.cpp

```diff
@@ -0,0 +1,192 @@
+
+  settings_manager_->SetBool(Param::fixed_bool, true, action_context, setter_callback);
+  EXPECT_EQ(common::ActionState::FAILURE, action_context->GetState());
+
+  settings_manager_->SetDouble(Param::fixed_double, 100.0, action_context, setter_callback);
+  EXPECT_EQ(common::ActionState::FAILURE, action_context->GetState());
+
+  settings_manager_->SetString(Param::fixed_string, "abcdefg", action_context, setter_callback);
+  EXPECT_EQ(common::ActionState::FAILURE, action_context->GetState());
+
+  // Test tunable parameters.
+  settings_manager_->SetInt(Param::lucky_number, 1919810, action_context, setter_callback);
+  EXPECT_EQ(common::ActionState::FAILURE, action_context->GetState());
+  EXPECT_EQ(114, settings_manager_->GetInt(Param::lucky_number));
+  settings_manager_->SetInt(Param::lucky_number, 233, action_context, setter_callback);
+  EXPECT_EQ(233, settings_manager_->GetInt(Param::lucky_number));
+
+  double pi = settings_manager_->GetDouble(Param::pi);
+  EXPECT_EQ(pi, 3.14159);
+  settings_manager_->SetDouble(Param::pi, 3.14, action_context, setter_callback);
+  EXPECT_EQ(3.14, settings_manager_->GetDouble(Param::pi));
+  settings_manager_->SetDouble(Param::pi, 0.0, action_context, setter_callback);
+  EXPECT_EQ(common::ActionState::FAILURE, action_context->GetState());
+
+  bool parallel = settings_manager_->GetBool(Param::parallel_execution);
+  EXPECT_TRUE(parallel);
+  settings_manager_->SetBool(Param::parallel_execution, false, action_context, setter_callback);
+  EXPECT_FALSE(settings_manager_->GetBool(Param::parallel_execution));
+
+  std::string_view name = settings_manager_->GetString(Param::db_name);
+  EXPECT_EQ("Terrier", name);
+  settings_manager_->SetString(Param::db_name, "PelotonSP", action_context, setter_callback);
+  EXPECT_EQ("PelotonSP", settings_manager_->GetString(Param::db_name));
+}
+
+// NOLINTNEXTLINE
+TEST_F(SettingsTests, CallbackTest) {
+  auto bufferPoolSize = static_cast<int64_t>(settings_manager_->GetInt(Param::buffer_pool_size));
+  EXPECT_EQ(bufferPoolSize, defaultBufferPoolSize);
+
+  bufferPoolSize = txn_manager_->GetBufferPoolSizeLimit();
+  EXPECT_EQ(bufferPoolSize, defaultBufferPoolSize);
+
+  const int32_t action_id = 1;
+  setter_callback_fn setter_callback = SettingsTests::EmptySetterCallback;
+  std::shared_ptr<common::ActionContext> action_context = std::make_shared<common::ActionContext>(action_id);
+
+  // Setting new value should invoke callback.
+  const int64_t newBufferPoolSize = defaultBufferPoolSize + 1;
+  settings_manager_->SetInt(Param::buffer_pool_size, static_cast<int32_t>(newBufferPoolSize), action_context,
+                            setter_callback);
+  bufferPoolSize = static_cast<int64_t>(settings_manager_->GetInt(Param::buffer_pool_size));
+  EXPECT_EQ(bufferPoolSize, newBufferPoolSize);
+
+  bufferPoolSize = txn_manager_->GetBufferPoolSizeLimit();
+  EXPECT_EQ(bufferPoolSize, newBufferPoolSize);
+}
+
+// NOLINTNEXTLINE
+TEST_F(SettingsTests, ConcurrentModifyTest) {
```

Comment:
What if multiple transactions are modifying the same settings? I think it is better to add this test.